Add ```core.transfers```

mattromano opened the pull request:
# Transfers

- Built Schema based on suggestions from discussion on #3 and from discussions in the near discord channel

# Testing

## dbt run --select core.transfers
``` root@d3c3a115050c:/near# dbt run --select core.transfers
Running with dbt=0.21.1
Found 6 models, 59 tests, 0 snapshots, 0 analyses, 175 macros, 0 operations, 0 seed files, 2 sources, 0 exposures

17:42:50 | Concurrency: 4 threads (target='dev')
17:42:50 | 
17:42:50 | 1 of 1 START incremental model DEV.transfers......................... [RUN]
17:43:21 | 1 of 1 OK created incremental model DEV.transfers.................... [SUCCESS 1 in 31.88s]
17:43:21 | 
17:43:21 | Finished running 1 incremental model in 34.73s.

Completed successfully

Done. PASS=1 WARN=0 ERROR=0 SKIP=0 TOTAL=1
```

## dbt test
```
root@d3c3a115050c:/near# dbt test --select core.transfers
Running with dbt=0.21.1
Found 6 models, 58 tests, 0 snapshots, 0 analyses, 175 macros, 0 operations, 0 seed files, 2 sources, 0 exposures

18:09:55 | Concurrency: 4 threads (target='dev')
18:09:55 | 
18:09:55 | 1 of 8 START test not_null_transfers_block_timestamp................. [RUN]
18:09:55 | 2 of 8 START test not_null_transfers_gas_used........................ [RUN]
18:09:55 | 3 of 8 START test not_null_transfers_receipt......................... [RUN]
18:09:55 | 4 of 8 START test not_null_transfers_tx_fee.......................... [RUN]
18:09:58 | 1 of 8 PASS not_null_transfers_block_timestamp....................... [PASS in 2.84s]
18:09:58 | 5 of 8 START test not_null_transfers_tx_receiver..................... [RUN]
18:09:58 | 2 of 8 PASS not_null_transfers_gas_used.............................. [PASS in 2.85s]
18:09:58 | 6 of 8 START test not_null_transfers_tx_signer....................... [RUN]
18:09:58 | 4 of 8 PASS not_null_transfers_tx_fee................................ [PASS in 2.86s]
18:09:58 | 7 of 8 START test not_null_transfers_txn_hash........................ [RUN]
18:10:00 | 6 of 8 PASS not_null_transfers_tx_signer............................. [PASS in 1.93s]
18:10:00 | 8 of 8 START test unique_transfers_txn_hash.......................... [RUN]
18:10:00 | 7 of 8 PASS not_null_transfers_txn_hash.............................. [PASS in 1.93s]
18:10:00 | 5 of 8 PASS not_null_transfers_tx_receiver........................... [PASS in 1.97s]
18:10:00 | 3 of 8 PASS not_null_transfers_receipt............................... [PASS in 4.90s]
18:10:02 | 8 of 8 PASS unique_transfers_txn_hash................................ [PASS in 1.42s]
18:10:02 | 
18:10:02 | Finished running 8 tests in 8.13s.

Completed successfully

Done. PASS=8 WARN=0 ERROR=0 SKIP=0 TOTAL=8
```

forgxyz commented on models/core/transfers.sql:
```diff
@@ -0,0 +1,64 @@
+actions as (
+
+  select
+    tx_id as txn_hash,
```
There's an outstanding PR to correct the column in `stg_txs` to `txn_hash` so this will need to be altered to account for that coming change.

forgxyz commented on models/core/transfers.sql:
```diff
@@ -0,0 +1,64 @@
+    tx_fee,
+    gas_used,
+    status,
+    try_parse_json(receipt) as receipt
```
There's a receipts table issue, so would it make sense to include `receipt_id` instead of the full object, here?

forgxyz reviewed (commented):
Generally, looks good. I have 2 in-line comments for review.

chuxinh commented on models/core/transfers.sql:
```diff
@@ -0,0 +1,64 @@
+    materialized='incremental',
+    cluster_by='block_timestamp',
+    unique_key='tx_id',
+    tags=['transfers']
```
```suggestion
    tags=['core', 'transfers']
```

chuxinh commented on models/core/transfers.sql:
```diff
@@ -0,0 +1,64 @@
+
+  )
+  
+select * from final
```
add extra line below

chuxinh replied:
@mattromano don't forget to add an extra line for this one :D

mattromano replied:
added!

chuxinh commented on models/core/transfers.sql:
```diff
@@ -0,0 +1,64 @@
+        else 'Failed'
+    end as status
+  from txs, 
+  lateral flatten( input => tx:actions )
```
i'm wondering instead of doing another flattening from the original `tx` table, how about actually using @forgxyz table `action_events` to build `transfer`?

mattromano replied:
Yeah I'm happy to do that as well, I imagine lateral flatten is fairly taxing so would probably save some resources. Also should be pretty easy to switch to from stg_txs to transfers table.

mattromano replied:
Actually sorry, after looking back on this I am getting a lot of data from the reciepts data so we would either need to expand the scope of the actions table to include receipt data (tx_fee, gas_used etc.) or just create this table from the ```stg_txs```, or potentially join the transfers names so we won't have to do the lateral flatten twice.

forgxyz replied:
There's an open issue for the `receipts` table, so we could keep it as-is, or wait until `receipts` is done and pull from that table. Separately, let's talk on discord @mattromano about what you are using from the `receipts` object as that will help direct the schema for `receipts.

chuxinh replied:
oops i missed out the receipt part. we can either wait for the receipt or what i'm thinking is use `action_events` joined with `transaction` on `txn_hash` if anything else is needed. 

I think use `action_events` you can easily get `Transfer` together with `FunctionCall` with deposit values

mattromano replied:
Okay cool, I think I'll wait till we merge at least  the transaction table and then build off of that + action events. Then down the road I can Update once the receipt table is built as well.

chuxinh commented on models/core/transfers.sql:
```diff
@@ -0,0 +1,65 @@
+      else OBJECT_KEYS(value)[0]::string
+    end as action_name,
+    actions_data:Transfer.deposit::int as deposit,
+    receipt:outcome:tokens_burnt::int + tx:outcome:outcome:tokens_burnt::int as tx_fee,
+    receipt:outcome:gas_burnt::int + tx:outcome:outcome:gas_burnt::int as gas_used,
```
yea i think if you join the later to be merged `transactions` table, the calculation of gas and fee can be avoided too https://github.com/MetricsDAO/near_dbt/pull/26/files#diff-965d9db004ce5501684e009c9cc9ccaab13c46cc7feda282ae646e00a5b77e47R33-R41

mattromano commented:
I have updated the sql and yml to build from the ```actions_events``` table and the ```transactions``` table.

### Changes include:

- Added ```action_id``` as the new unique table key. Noticed there were a few transfers which had two deposit values under one ```action_data``` object so ```txn_hash``` wasn't actually unique
- Replaced ```receipt``` json object with ```receipt_id``` so we can reference future receipt table
- Replaced lateral flatten transformation by referencing ```actions_events``` table and the ```transactions``` table
- Updated all to reference ```transaction``` table instead of ```stg_txn```

## Tests
### dbt run --select core.transfers
```root@c601d2bd6928:/near# dbt run --select core.transfers
Running with dbt=0.21.1
Found 6 models, 61 tests, 0 snapshots, 0 analyses, 173 macros, 0 operations, 0 seed files, 2 sources, 0 exposures

02:28:15 | Concurrency: 4 threads (target='dev')
02:28:15 | 
02:28:15 | 1 of 1 START incremental model DEV.transfers......................... [RUN]
02:28:21 | 1 of 1 OK created incremental model DEV.transfers.................... [SUCCESS 1 in 6.25s]
02:28:21 | 
02:28:21 | Finished running 1 incremental model in 11.80s.
```
### dbt test --select core.transfers
```
root@c601d2bd6928:/near# dbt test --select core.transfers
Running with dbt=0.21.1
Found 6 models, 61 tests, 0 snapshots, 0 analyses, 173 macros, 0 operations, 0 seed files, 2 sources, 0 exposures

02:28:36 | Concurrency: 4 threads (target='dev')
02:28:36 | 
02:28:36 | 1 of 9 START test not_null_transfers_action_id....................... [RUN]
02:28:36 | 2 of 9 START test not_null_transfers_block_timestamp................. [RUN]
02:28:36 | 3 of 9 START test not_null_transfers_gas_used........................ [RUN]
02:28:36 | 4 of 9 START test not_null_transfers_receipt_id...................... [RUN]
02:28:39 | 1 of 9 PASS not_null_transfers_action_id............................. [PASS in 3.33s]
02:28:39 | 5 of 9 START test not_null_transfers_status.......................... [RUN]
02:28:39 | 4 of 9 PASS not_null_transfers_receipt_id............................ [PASS in 3.33s]
02:28:39 | 2 of 9 PASS not_null_transfers_block_timestamp....................... [PASS in 3.34s]
02:28:39 | 3 of 9 PASS not_null_transfers_gas_used.............................. [PASS in 3.34s]
02:28:39 | 6 of 9 START test not_null_transfers_transaction_fee................. [RUN]
02:28:39 | 7 of 9 START test not_null_transfers_tx_receiver..................... [RUN]
02:28:39 | 8 of 9 START test not_null_transfers_tx_signer....................... [RUN]
02:28:42 | 7 of 9 PASS not_null_transfers_tx_receiver........................... [PASS in 2.80s]
02:28:42 | 9 of 9 START test unique_transfers_action_id......................... [RUN]
02:28:42 | 5 of 9 PASS not_null_transfers_status................................ [PASS in 2.90s]
02:28:42 | 6 of 9 PASS not_null_transfers_transaction_fee....................... [PASS in 2.88s]
02:28:42 | 8 of 9 PASS not_null_transfers_tx_signer............................. [PASS in 3.00s]
02:28:44 | 9 of 9 PASS unique_transfers_action_id............................... [PASS in 1.70s]
02:28:44 | 
02:28:44 | Finished running 9 tests in 10.33s.

Completed successfully

Done. PASS=9 WARN=0 ERROR=0 SKIP=0 TOTAL=9
```

chuxinh commented on models/core/transfers.sql:
```diff
@@ -0,0 +1,60 @@
+    a.deposit,
+    t.transaction_fee,
+    t.gas_used,
+    t.tx_receipt[0]:id::string as receipt_id,
```
there's a few receipt ids there, how to identify that the first one is related to this action transfer?

mattromano replied:
So as I understand it there looks like there is an id to the actual receipt object, and then underneath that there is a separate portion called ```receipt_ids```, which in the case of a failed txn there are two  ```receipt_ids``` or if it is successful it is one ```receipt_id```. Both of which are different from the actual receipt object id. 

I think we originally wanted this field to be able to link back to the the receipt object once the receipt table has been parsed out. I think maybe we could just re-name the field to be ```reciept_object_id``` as to not get confused with the actual ```reciept_ids``` that are in the receipt object. Sorry, hopefully that makes sense, also if I am completely wrong here let me know haha

chuxinh replied:
that makes sense to me! 
do you know if this `reciept_object_id` will be the primary key to join the `receipt` table?

mattromano replied:
@forgxyz 
Looking into receipts a bit further here and it looks like there's a ```receipt_ids``` field in the actual receipt object (as well as in the outcome object), separate from the ```id``` field at the top (which for clarification's sake I am considering is the ```receipt_object_id``` in this model/query). 

It also looks like the outcome object's `reciept_ids` always = ```receipt_object_id``` . I just checked and there is not a case where outcome's ```receipt_ids``` did not equal ```receipt_object_id```.  I am still not sure if  or what the receipt object's ```receipt_ids``` link to  other than the fact we get two of them if the transaction fails. 

All thathat being said, do you think ```receipt_object_id``` would make sense as a primary key for the receipt table? Want to confirm before we merge so we can make this table is joinable to `receipt` table in the future.

forgxyz replied:
Ok, reading the thread here I agree that it makes sense to use `receipt_object_id` and that will be the PK of the pending `receipts` table. @mattromano I think we are good to merge if this was the last blocker.

chuxinh commented on models/core/transfers.sql:
```diff
@@ -0,0 +1,60 @@
+    t.tx_receipt[0]:id::string as receipt_id,
+    case
+        when tx_receipt[0]:outcome:status::string = '{"SuccessValue":""}' then 'Succeeded' 
+        else 'Failed'
```
or maybe turn this into a boolean?

mattromano replied:
Yeah I can do that, it will make it easier on storage. Only thought was to follow the block explorer but I think we can include that True corresponds to "Succeeded" and "Failed" corresponds to False within the yml docs.

chuxinh commented on models/core/transfers.yml:
```diff
@@ -0,0 +1,58 @@
+          - not_null
+
+      - name: deposit 
+        description: Sum of all NEAR tokens transferred from the Signing account to the Receiver account. This includes tokens sent in a Transfer action(s), and as deposits on Function Call action(s). In raw number format, to adjust divide by POW(10,24)
```
does it include function call action deposits too?

mattromano replied:
Oh no it doesn't I just copied from NEAR block explorer and forgot to delete that portion. I'll delete that portion and keep it transfers specific.

chuxinh reviewed (approved):
LGTM after nits
